src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -81,37 +81,447 @@ constexpr const ad_utility::Last<Current, Others...>* unwrapVariant(
+
+// ___________________________________________________________________________
+ObjectList SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::ObjectListPathContext* ctx) {
+  // The second parameter is empty because collections and blank not paths,
+  // which might add additional triples, are currently not supported.
+  return ObjectList{visitVector<VarOrTerm>(ctx->objectPath()), {}};
+}
+
+// ___________________________________________________________________________
+PropertyPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PathContext* ctx) {
+  return visitTypesafe(ctx->pathAlternative());
+}
+
+// ___________________________________________________________________________
+PropertyPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PathAlternativeContext* ctx) {
+  auto paths = visitVector<PropertyPath>(ctx->pathSequence());
+
+  if (paths.size() == 1) {
+    return paths[0];
+  } else {
+    return PropertyPath::makeAlternative(std::move(paths));
+  }
+}
+
+// ___________________________________________________________________________
+PropertyPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PathSequenceContext* ctx) {
+  auto paths = visitVector<PropertyPath>(ctx->pathEltOrInverse());
+
+  if (paths.size() == 1) {
+    return paths[0];
+  } else {
+    return PropertyPath::makeSequence(std::move(paths));
+  }
+}
+
+// ___________________________________________________________________________
+ad_utility::sparql_types::VarOrPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::VerbPathOrSimpleContext* ctx) {
+  if (ctx->verbPath()) {
+    return visitTypesafe(ctx->verbPath());
+  } else if (ctx->verbSimple()) {
+    return visitTypesafe(ctx->verbSimple());
+  }
+  AD_FAIL()  // Should be unreachable.
+}
+
+// ___________________________________________________________________________
 PathTuples SparqlQleverVisitor::visitTypesafe(
     SparqlAutomaticParser::PropertyListPathNotEmptyContext* ctx) {
+  PathTuples tuples = visitTypesafe(ctx->tupleWithPath());
+  vector<PathTuples> tuplesWithoutPaths =
+      visitVector<PathTuples>(ctx->tupleWithoutPath());
+  for (auto& tuplesWithoutPath : tuplesWithoutPaths) {
+    tuples.insert(tuples.end(), tuplesWithoutPath.begin(),
+                  tuplesWithoutPath.end());
+  }
+  return tuples;
+}
+
+// ___________________________________________________________________________
+PropertyPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PathEltContext* ctx) {
+  PropertyPath p = visitTypesafe(ctx->pathPrimary());
+
+  if (ctx->pathMod()) {
+    if (ctx->pathMod()->getText() == "+") {
+      p = PropertyPath::makeTransitiveMin(p, 1);
+    } else if (ctx->pathMod()->getText() == "?") {
+      p = PropertyPath::makeTransitiveMax(p, 1);
+    } else if (ctx->pathMod()->getText() == "*") {
+      p = PropertyPath::makeTransitive(p);
+    } else {
+      AD_FAIL()  // Should be unreachable.
+    }
+  }
```

Comment:
Maybe we can even make this typesafe by having an enum for the PathMods which can then be visited, but we can discuss this.